File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] l = new int[n];
        l[0] = 1;
        int[] r = new int[n];
        r[n-1] = 1;
        for(int i=1;i<n;i++){
            l[i] = l[i-1]*nums[i-1];
            r[n-i-1] = r[n-i]*nums[n-i];
        }
        int[] res = new int[n];
        for(int i=0;i<n;i++){
            res[i] = l[i]*r[i];
        }
        return res;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        lt = [1]*n
        res = [1] *n
        for i in range(1, n):
            lt[i] = nums[i-1]*lt[i-1]
        rt = 1
        for i in range(n-1, -1, -1):
            res[i] = lt[i] * rt
            rt = rt * nums[i]
        return res
        
            
        